### Import Libraries

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [3]:
# loading the data set
# Credit Crad Data - ccd
ccd = pd.read_csv('creditcard.csv')

In [4]:
ccd

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

* Time - first transaction at 0th second, and other are relative to it
* V1,V2 are different variables related to credit card and for security purpose they are given under such labels
* Amount - Transaction amount in dollars
* Class - 0 Legit, 1 Fradulent

In [5]:
# Dataset information
ccd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [6]:
# Checking for missing values in Dataset
ccd.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [7]:
# Checking the distribution of legit transactions and fradulent transactions
ccd['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

* 0 ---> Normal Transaction
* 1 ---> Fraudulent Transaction

In [8]:
Fraudulent_Transaction_Percentage = (492/284315)*100 
print(Fraudulent_Transaction_Percentage)

0.17304750013189596


only 0.17% data is fraudulent which is
very less

* This is highly unbalanced dataset
* As the no of fradulent records are very less
* ML will always predict a legit transaction

### Seperating the data for Analysis

In [9]:
legit = ccd[ccd.Class==0]
fraud = ccd[ccd.Class==1]

In [10]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


### Statistical measures of the data

#### Amount of Transaction Comparison

In [11]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [12]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

* We can clearly see the difference between means and quantile values

### Comparing the mean of values of both transactions

In [13]:
ccd.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

### Dealing with Unbalanced Data

* Technique - Under-sampling
* Build a sample dataset containing similar distribution of normal and fraudulent transactions
* i.e. we have 492 fraud transaction, so we will choose 492 normal transactions from 284315 randomly and then develop the ML model.

#### Taking a random sample of size 492 from normal transactions

In [14]:
legit_sample = legit.sample(n=492)

#### Now, concat this sample with fraud data to make a new dataframe which will be ready to analyze

In [15]:
new_ccd = pd.concat([legit_sample,fraud],axis=0)

In [16]:
new_ccd

Time        V1        V2        V3        V4        V5        V6  \
184587  126301.0  2.260904 -0.325432 -2.342071 -0.387530  0.115165 -1.852174   
245290  152682.0  2.133566 -0.080722 -2.285663 -0.634283  0.837700 -0.744619   
209579  137625.0  1.961343 -0.715029 -1.703147 -0.613582  1.713865  3.913136   
224823  143978.0  1.349577 -2.101880 -3.326968 -0.644385  0.082312 -1.016522   
53734    46122.0  1.316315 -1.513755  0.088182 -1.534671 -1.225085  0.070248   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
184587  0.585789 -0.702393 -1.137272  ... -0.106839  0.343126 -0.085222   
245290  0.506136 -0.327503 -0.192638  ...  0.078304  0.264425  0.034851   
209579 -1.252671  1.054911  1.314417  ...  0.191787  0.691780  0.158595   
224823  1.063473 -0.674772 -1.016286  ...  0.748940  1.043497 -0.684967   
53734  -1.030388 -0.138505 -1.649654  ... -0.054249 -0.022394 -0.312133   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
184587  0.116794  0.499055  1.017775 -0.122012 -0.092666    6.00      0  
245290  0.332895  0.122700  1.326961 -0.176528 -0.098408    9.99      0  
209579  0.726621 -0.060118 -0.192346  0.060914 -0.042846    9.99      0  
224823  0.381692  0.625843  0.245364 -0.197234 -0.003911  519.70      0  
53734  -0.980699  0.503571 -0.035399  0.028223  0.040429  158.00      0  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[984 rows x 31 columns]

In [17]:
# Checkout the distribution again!!
new_ccd['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [18]:
# Now again check and compare the means of both data sets
new_ccd.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94283.780488  0.051233 -0.145487 -0.020955 -0.062594  0.071685   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.041384  0.001011 -0.136335 -0.007725  ...  0.016713  0.088111   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.045628  0.004298  0.029528 -0.009173  0.004885  0.007728 -0.000569   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0      103.208191  
1      122.211321  

[2 rows x 30 columns]

* Means value are nearly same for sample also
* Which reflects that the sample is correctly representing the population
* i.e. the nature of the dataset hasn't changed

### Seperate the Dependent and Independent Variables

* Now, seperate the Dependent and Independent variables
* Dependent Variable (y) - Class
* Independent variable (x) - All variables except class

In [22]:
x = new_ccd.drop(columns = 'Class',axis=1)
y = new_ccd['Class']

In [23]:
print(x)

            Time        V1        V2        V3        V4        V5        V6  \
184587  126301.0  2.260904 -0.325432 -2.342071 -0.387530  0.115165 -1.852174   
245290  152682.0  2.133566 -0.080722 -2.285663 -0.634283  0.837700 -0.744619   
209579  137625.0  1.961343 -0.715029 -1.703147 -0.613582  1.713865  3.913136   
224823  143978.0  1.349577 -2.101880 -3.326968 -0.644385  0.082312 -1.016522   
53734    46122.0  1.316315 -1.513755  0.088182 -1.534671 -1.225085  0.070248   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [24]:
print(y)

184587    0
245290    0
209579    0
224823    0
53734     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


### Now Split the data into training and testing dataset

In [25]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=2, stratify=y)

In [26]:
print(x.shape, x_train.shape, x_test.shape)

(984, 30) (787, 30) (197, 30)


In [27]:
print(y.shape, y_train.shape, y_test.shape)

(984,) (787,) (197,)


### Model Fitting

* Logistic Regression
* Since the Class variable has binary output, i.e. 0 and 1

In [31]:
# Training the logistic regression model with training dataset
model = LogisticRegression() # saving the logistic regression model in a model variable for ease of accessibility

In [32]:
model.fit(x_train, y_train)

LogisticRegression()

### Model Evolution by Accuracy Score

#### accuracy on training dataset

In [36]:
x_train_prediction = model.predict(x_train) # i.e. predictions for corresponding y variable

In [37]:
acc_scr = accuracy_score(x_train_prediction, y_train)

In [40]:
print('Accuracy Score on training dataset :',acc_scr)

Accuracy Score on training dataset : 0.9351969504447268


* Good Accuracy score obtained
* Accuracy Score greater than 75 or 80 is considered as a Good Score

### Note

* But accuracy score on testing dataset is most important because we want to see how our model performs on unknown data
* As our model has not seen the test dataset

#### accuracy on testing dataset

In [41]:
x_test_prediction = model.predict(x_test)

In [42]:
acc_scr1=accuracy_score(x_test_prediction, y_test)

In [43]:
print('Accuracy Score on testing dataset :',acc_scr1)

Accuracy Score on testing dataset : 0.9289340101522843


* This is also a good accuracy score

### Note

* Accuracy score for training dataset and testing dataset is very close
* This tells us that the model is fitting good
* If there is significant difference between accuracy scores of training dataset and testing dataset, the the model either over-fitting or under-fitting

### Interpretation - 

* If accuracy score for training dataset say, 94% and that is for testing dataset is 50 to 60%, then the model will be overfitting
* Because the model is fiiting really well on training dataset, but failed to perform on testing dataset
* Conversly, if testing data accuracy is very high as compared to training data, the model is considered as underfitting
* We don't want both